libraries/bot-builder/src/main/java/com/microsoft/bot/builder/skills/SkillHandlerImpl.java

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.bot.builder.skills;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.microsoft.bot.builder.Bot;
+import com.microsoft.bot.builder.BotAdapter;
+import com.microsoft.bot.builder.BotCallbackHandler;
+import com.microsoft.bot.builder.TurnContext;
+import com.microsoft.bot.connector.authentication.ClaimsIdentity;
+import com.microsoft.bot.connector.authentication.JwtTokenValidation;
+import com.microsoft.bot.schema.Activity;
+import com.microsoft.bot.schema.ActivityTypes;
+import com.microsoft.bot.schema.CallerIdConstants;
+import com.microsoft.bot.schema.ConversationReference;
+import com.microsoft.bot.schema.ResourceResponse;
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+/**
+ * This class inherited all the implementations of {@link SkillHandler} class since we needed similar code for {@link CloudSkillHandler}.
+ * The {@link CloudSkillHandler} class differs from {@link SkillHandler} class only in authentication by making use of {@link BotFrameworkAuthentication} class.
+ * This class is internal since it is only used in skill handler classes.
+ */
+public class SkillHandlerImpl {
+
+    private final String skillConversationReferenceKey;
+    private final BotAdapter adapter;
+    private final Bot bot;
+    private final SkillConversationIdFactoryBase conversationIdFactory;
+    private final Supplier<String> getOAuthScope;
+    private Logger logger = LoggerFactory.getLogger(SkillHandlerImpl.class);
+
+    public SkillHandlerImpl(
+        String withSkillConversationReferenceKey,
+        BotAdapter withAdapter,
+        Bot withBot,
+        SkillConversationIdFactoryBase withConversationIdFactory,
+        Supplier<String> withGetOAuthScope) {
+        if (StringUtils.isBlank(withSkillConversationReferenceKey)) {
+            throw new IllegalArgumentException("skillConversationReferenceKey cannot be null");
+        }
+        if (withAdapter == null) {
+            throw new IllegalArgumentException("adapter cannot be null");
+        }
+
+        if (withBot == null) {
+            throw new IllegalArgumentException("bot cannot be null");
+        }
+
+        this.skillConversationReferenceKey = withSkillConversationReferenceKey;
+        this.adapter = withAdapter;
+        this.bot = withBot;
+        this.conversationIdFactory = withConversationIdFactory;
```

Comment:
conversationIdFactory and getOAuthScope not being [checked for null](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder/Skills/SkillHandlerImpl.cs#L47)?

Reply:
Thanks, fixed!